### PR TITLE
Implement blue/green deployment support in REPL mode

### PR DESCRIPTION
**Pull request by kaituo**
### Description
Features:
- Add mutual exclusivity in session-based emr-s jobs to ensure only one job runs at a time, enhancing system stability during blue/green deployments. This allows active job exclusion and seamless task pickup by new jobs. Details in #94.

Fixes:
- Resolve a bug where long-running queries failed to cancel post-index mapping verification, by introducing timely query cancellation checks within the REPL loop.
- Address issue #138 with the proposed short-term fix, improving reliability.

Tests:
- Conducted manual testing to validate blue/green deployment support and query cancellation.
- Extended unit tests to cover new features and bug fixes.

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/138
https://github.com/opensearch-project/opensearch-spark/issues/94

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala [hunk 1]
```diff
@@ -5,21 +5,30 @@
...
 
-import org.apache.spark.SparkFunSuite
+import org.apache.spark.{SparkContext, SparkFunSuite}
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.catalyst.trees.Origin
 import org.apache.spark.sql.types.{ArrayType, LongType, NullType, StringType, StructField, StructType}
 import org.apache.spark.sql.util.ShutdownHookManagerTrait
+import org.apache.spark.util.ThreadUtils
 
 class FlintREPLTest
```
Comment (penghuo) [inline spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala:33]: if statement run successfully, but write result failed. what is the expectation?
1. update statementState = FAILED. error = "write result failed"

Comment (penghuo) [reply spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala:33]: Try to understand the failure model if OpenSearch indices is unavailable? and how can we can fail statement and job properly.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 2]
```diff
@@ -83,27 +86,32 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
-          resultIndex,
-          dataSource,
-          sessionIndex.get,
-          sessionId.get,
-          spark,
-          osClient,
-          jobId,
-          applicationId,
-          flintSessionIndexUpdater)
+        exponentialBackoffRetry(maxRetries = 5, initialDelay = 2.seconds) {
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:114]: nice! it cover transient OpenSearch cluster index not available, include request_index and result index, right?
* if request_index not available, retry 5 times, and update session state = fail, job exit. but in case request_index is not available, spark job can not update session state, so session state could be `not_started` or `running`.  if it is not_start, no problem. if it is running, plugin should handle it properly. https://github.com/opensearch-project/sql/issues/2415
* if result_index not available, how do we handle the failed statement?

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 3]
```diff
@@ -634,4 +693,93 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
+        val source = getResponse.getSourceAsMap
+        if (source == null) {
+          logError(s"""Session id ${sessionId} is empty""")
+          // still proceed since we are not sure what happened (e.g., OpenSearch cluster may be unresponsive)
+          return true
+        }
+
+        val runJobId = Option(source.get("jobId")).map(_.asInstanceOf[String]).orNull
+        val excludeJobIds: Seq[String] = {
+          val rawExcludeJobIds = source.get("excludeJobIds")
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:757]: i did not found code store excludeJobIds in OpenSearch session doc.

File: flint-spark-integration/src/main/scala/org/opensearch/flint/app/FlintInstance.scala [hunk 4]
```diff
@@ -19,6 +18,8 @@ class FlintInstance(
     val sessionId: String,
     val state: String,
     val lastUpdateTime: Long,
+    // We need jobStartTime to check if HMAC token is expired or not
```
Comment (penghuo) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/app/FlintInstance.scala:21]: remove this comments. 
We need jobStartTime to log the when does job started.

File: spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala [hunk 5]
```diff
@@ -170,4 +179,396 @@ class FlintREPLTest
...
+        dataSource,
+        sessionId,
+        executionContext,
+        startTime,
+        // make sure it times out before mockSparkSession.sql can return, which takes 60 seconds
+        Duration(1, SECONDS))
+
+      verify(mockSparkSession, times(1)).sql(any[String])
+      verify(sparkContext, times(1)).cancelJobGroup(any[String])
+      result should not be None
```
[the remark this answers is not in the document]
Comment (kaituo) [reply spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala:502]: it is a dataframe with error info

File: spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala [hunk 6]
```diff
@@ -170,4 +179,396 @@ class FlintREPLTest
...
+        sessionId,
+        executionContext,
+        startTime,
+        Duration.Inf // Use Duration.Inf or a large enough duration to avoid a timeout
+      )
+
+      // Verify that ParseException was caught and handled
+      result should not be None // or result.isDefined shouldBe true
+      flintCommand.error should not be None
+      flintCommand.error.get should include("Syntax error:")
```
[the remark this answers is not in the document]
Comment (kaituo) [reply spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala:552]: yes, sth like

"error": """Syntax error: 
Syntax error at or near '1': extra input '1'(line 1, pos 39)

== SQL ==
select * from default.http_logs limit1 1
---------------------------------------^^^
"""

File: spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala (hunk 1 above)
Comment (kaituo) [reply spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala:33]: > if statement run successfully, but write result failed. what is the expectation?

    update statementState = FAILED. error = "write result failed"

Yes. That's the expectation. Error will contain more details.

This is the error handling when write result failed.

```
      case e: Exception =>
        val error = s"""Fail to write result of ${flintCommand}, cause: ${e.getMessage}"""
```

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 7]
```diff
@@ -83,92 +86,120 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
+          HEARTBEAT_INTERVAL_MILLIS,
+          flintSessionIndexUpdater,
           sessionId.get,
-          spark,
+          threadPool,
           osClient,
-          jobId,
+          sessionIndex.get)
+
+        setupFlintJob(
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:106]: we should update excludeJobIds in here.
* read excludeJobIds from --conf spark.flint.deployment.excludeJobs = "job-1", the configure is optional.
* update FlintJob metadata
  * read existing FlintJob before write, get (jobId, excludeJobs)
  * update FlintJob only when
```
confExcludeJobs = conf.get("spark.flint.deployment.excludeJobs")
if (confExcludeJobs == null) {
  updateFlintJob()
} else {
   if (excludeJobs include currentJobId) // edge case, current job is excluded.
  { exit; }
   if (confExcludeJobs == excludeJobs) // edge case, duplicate job running
  { exit; }
}
```

File: spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala (hunk 1 above)
Comment (kaituo) [reply spark-sql-application/src/test/scala/org/apache/spark/sql/FlintREPLTest.scala:33]: When OpenSearch indices are not available, we will fail the session and write the failed session to OpenSearch indices and terminate the job. But we won't fail statements as we cannot read from OpenSearch indices.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 2 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:114]: If result_index not available because connection to OpenSearch is lost, I added logic to check if result index exists or not. If it doesn't, recreate one with correct mapping.

Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:114]: Yes, your understanding of request_index not available is right.

File: flint-spark-integration/src/main/scala/org/opensearch/flint/app/FlintInstance.scala (hunk 4 above)
Comment (kaituo) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/app/FlintInstance.scala:21]: removed

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 3 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:757]: will read it from spark properties

**Review by penghuo — COMMENTED** [review]
could we make following value configurable. i found with 2 executors, some query can not finish by 10mins, but just in case user want to override it.

private val INACTIVITY_LIMIT_MILLIS = 30 * 60 * 1000
private val QUERY_EXECUTION_TIMEOUT = Duration(10, MINUTES)
private val QUERY_WAIT_TIMEOUT_MILLIS = 10 * 60 * 1000

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 7 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala]: implemented

**Comment by kaituo** [comment]
> could we make following value configurable. i found with 2 executors, some query can not finish by 10mins, but just in case user want to override it.
> 
> private val INACTIVITY_LIMIT_MILLIS = 30 * 60 * 1000 private val QUERY_EXECUTION_TIMEOUT = Duration(10, MINUTES) private val QUERY_WAIT_TIMEOUT_MILLIS = 10 * 60 * 1000

changed.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 8]
```diff
@@ -80,95 +84,221 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
+          flintSessionIndexUpdater,
+          sessionIndex.get,
+          jobStartTime)
+
+      case Some(confExcludeJobs) =>
+        // example: --conf spark.flint.deployment.excludeJobs=job-1,job-2
+        val excludeJobIds = confExcludeJobs.split(",").toList // Convert Array to Lis
+
+        if (excludeJobIds.contains(jobId)) {
+          // Edge case, current job is excluded, exit the application
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:227]: add log why job exit.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 9]
```diff
@@ -80,95 +84,221 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
+        }
+
+        val getResponse = osClient.getDoc(sessionIndex.get, sessionId.get)
+        if (getResponse.isExists()) {
+          val source = getResponse.getSourceAsMap
+          if (source != null) {
+            val existingExcludedJobIds = parseExcludedJobIds(source)
+            if (excludeJobIds.sorted == existingExcludedJobIds.sorted) {
+              // Edge case, duplicate job running, exit the application
+              return true
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:238]: add log why job exit.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 10]
```diff
@@ -607,16 +841,11 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
-              val flintInstant: FlintInstance = new FlintInstance(
-                source.get("applicationId").asInstanceOf[String],
-                source.get("jobId").asInstanceOf[String],
-                source.get("sessionId").asInstanceOf[String],
-                "running",
-                source.get("lastUpdateTime").asInstanceOf[Long],
-                Option(source.get("error").asInstanceOf[String]))
+              val flintInstance = FlintInstance.deserializeFromMap(source)
+              flintInstance.state = "running"
               flintSessionUpdater.updateIf(
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:846]: We should handle VersionConflictException with backoff retry.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 11]
```diff
@@ -264,43 +445,50 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
-      dataSource: String,
-      resultIndex: String,
-      sessionId: String,
-      futureMappingCheck: Future[Either[String, Unit]],
-      recordedVerificationResult: VerificationResult,
-      executionContext: ExecutionContextExecutor,
-      flintSessionIndexUpdater: OpenSearchUpdater,
-      recordedLastActivityTime: Long): (Long, VerificationResult) = {
+      context: CommandContext,
+      state: CommandState): (Long, VerificationResult, Boolean) = {
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:449]: does state been used in this function?

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 12]
```diff
@@ -508,22 +746,23 @@ object FlintREPL extends Logging with FlintJobExecutor {
...
-         |        "term": {
-         |          "dataSourceName": "$dataSource"
+         |        "range": {
+         |          "submitTime": { "gte": "now-1h" }
          |        }
          |      }
          |    ]
          |  }
          |}""".stripMargin
+
```
Comment (penghuo) [inline spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:765]: 1. state should be `waiting`. 
2. what if plugin cancel statement, but queryLoop use scroll search, the statement in scroll context still waiting state.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 8 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:227]: added

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 9 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:238]: added

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 10 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:846]: The code is related to heart beat update and we will log the VersionConflictException. Backoff retry is related to query execution. These two parts do not seem coincide.

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 11 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:449]: yes. The value is imported using

import state._

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala (hunk 12 above)
Comment (kaituo) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:765]: 1. Can you explain a bit? Are you asking where I search for waiting state

```
        "term": {
          "state": "waiting"
        }
      },
```

2. Yeah, we need to change to not to use scroll search.

**Review by penghuo — APPROVED** [review]
